### PR TITLE
[Intl] Update timezones to ICU 64.2 + compile zone to country mapping

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no     
| Deprecations? | no
| Tests pass?   | yes    
| Fixed tickets | #...   
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This PR compiles the zone to country mapping (and vice versa) from ICU data:
https://github.com/unicode-org/icu/blob/master/icu4c/source/data/misc/windowsZones.txt

I've recompiled timezones on master due #31162, i should do it once more when it's merged upstream.

Having this data allows compatibility between PHP timezones and ICU;
- https://www.php.net/manual/en/intltimezone.getregion.php
- https://www.php.net/manual/en/class.datetimezone.php#datetimezone.constants.per-country

For the timezone validator in Symfony, this would be required to have a compatible "regions" option, once it supports ICU as well (https://github.com/symfony/symfony/issues/28836#issuecomment-483769029)

### CODE REVIEW

**Comment by jakzal** [comment]
Nice, I was waiting for the 3.4 merge to master.

Make sure `intl-data` tests are passing please :) They're not being run on the CI.

**Comment by ro0NL** [comment]
this ones ready, see last commit for the added timezones which are missing currently (cc @fabpot)

cc @sstok tests are passing :+1: 

The main change is we use `Country Time (Location)` instead of `Location Time`

Also a fallback was missing for e.g. `zh_Hant_HK > zh_Hant > zh > root` instead of `zh_Hant_HK > root`. Given we build a timezone name from individual parts, we need the fallback chain as such during generation.

I think this nails it, and im done with timezones for a while now :laughing:

**Comment by fabpot** [comment]
@ro0NL Is it desired that we have a trailing white space in `src/Symfony/Component/Intl/Resources/data/timezones/*.json` files?

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php [hunk 1]
```diff
@@ -96,20 +121,63 @@ protected function generateDataForMeta(BundleReaderInterface $reader, $tempDir)
...
 
+        $data['CountryToZone'] = self::generateCountryToZoneMapping($data['ZoneToCountry']);
+
         return $data;
     }
 
-    private static function generateZones(ArrayAccessibleResourceBundle $typeBundle, ArrayAccessibleResourceBundle $metaBundle, ArrayAccessibleResourceBundle $rootBundle, ArrayAccessibleResourceBundle $localeBundle): array
+    private function generateZones(string $locale, ArrayAccessibleResourceBundle $typeBundle, ArrayAccessibleResourceBundle $metaBundle, ArrayAccessibleResourceBundle $windowsZonesBundle, ArrayAccessibleResourceBundle ...$localeBundles): array
     {
+        $accessor = function (ArrayAccessibleResourceBundle $resourceBundle, array $indices) {
```
Comment (sstok) [inline src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: This doesn't use the `$this` scope and can be made `static`.

**Review by sstok — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php (hunk 1 above)
Comment (ro0NL) [reply src/Symfony/Component/Intl/Data/Generator/TimezoneDataGenerator.php]: fixed.

**Comment by ro0NL** [comment]
@fabpot not sure :) all json files (en.json, composer.json) have trailing EOL; there are no parsing issues AFAIK. Unless you meant a different whitespace?

If it's about e.g. https://github.com/symfony/symfony/pull/31262/files#diff-5996b6b88384d2b755c17f069229882fR4 that's part of the localized interpunction; it's UTF :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ro0NL.

**Pull request merged**